src/librustc_mir/transform/const_prop.rs

```diff
@@ -74,6 +75,46 @@ impl<'tcx> MirPass<'tcx> for ConstProp {
             return;
         }
 
+        // Check if it's even possible to satisfy the 'where' clauses
+        // for this item.
+        // This branch will never be taken for any normal function.
+        // However, it's possible to `#!feature(trivial_bounds)]` to write
+        // a function with impossible to satisfy clauses, e.g.:
+        // `fn foo() where String: Copy {}`
+        //
+        // We don't usually need to worry about this kind of case,
+        // since we would get a compilation error if the user tried
+        // to call it. However, since we can do const propagation
+        // even without any calls to the function, we need to make
+        // sure that it even makes sense to try to evaluate the body.
+        // If there are unsatisfiable where clauses, then all bets are
+        // off, and we just give up.
+        //
+        // Note that we use TraitQueryMode::Canonical here, which causes
+        // us to treat overflow like any other error. This is because we
+        // are "speculatively" evaluating this item with the default substs.
+        // While this usually succeeds, it may fail with tricky impls
+        // (e.g. the typenum crate). Const-propagation is fundamentally
+        // "best-effort", and does not affect correctness in any way.
+        // Therefore, it's perfectly fine to just "give up" if we're
+        // unable to check the bounds with the default substs.
+        //
+        // False negatives (failing to run const-prop on something when we actually
+        // could) are fine. However, false positives (running const-prop on
+        // an item with unsatisfiable bounds) can lead to us generating invalid
+        // MIR.
+        if !tcx.substitute_normalize_and_test_predicates((
```

Comment:
If the substs are empty, can we use `TraitQueryMode::Standard`? Doing so should ensure that `substitute_normalize_and_test_predicates` isn't invoked twice when we know that the result will be the same irrelevant of the `TraitQueryMode`, right?

Reply:
When you say "invoked twice", do you mean `substitute_normalize_and_test_predicates` being invoked from somewhere else with the same `DefId` and `Substs`, but a different `TraitQueryMode`?

I think we could still have overflow with some really weird code (e.g a macro-generated `impl Foo for A where B: Foo`, `impl Foo for B where C: Foo`, etc). I'm not certain if we would end up with a "legitimate" overflow in that case (i.e. caused by something other than const-prop).

I think it's better to always use canonical mode here, as emitting an error here is always wrong, and could be very misleading to users (they might assume that their code has an issue, when it's really fine).

Reply:
Ah, that makes sense. Thanks